[JENKINS-54459] Avoid Angry Jenkins when check now on Plugin Manager

MRamonLeon opened the pull request:
See [JENKINS-54459](https://issues.jenkins-ci.org/browse/JENKINS-54459).

<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Proposed changelog entries

* Avoid Angry Jenkins in Plugin Manager - Check now button
* Add system properties to retry checking the update site: `hudson.PluginManager.sleepTimeMilis` (1000), `hudson.PluginManager.retries` (3)

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [ ] N/A For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@jtnord @reviewbybees

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

New UI after an error:
![screenshot from 2018-11-05 14-06-24](https://user-images.githubusercontent.com/22069922/47999850-0aac8f80-e104-11e8-95e6-1852cf6043c3.png)

Console:
![screenshot from 2018-11-05 10-19-52](https://user-images.githubusercontent.com/22069922/47999305-7261db00-e102-11e8-94ee-fa0a12006b53.png)

MRamonLeon committed:
[JENKINS-54459] Avoid angry Jenkins in Plugin Manager. Check Now

MRamonLeon committed:
[JENKINS-54459] Avoid angry Jenkins in Plugin Manager. Check Now

jtnord commented:
Code looks ok to me - but the underlying issue where the UpdateSite throws an exception rather than returning a `FormValidation.Error` looks like it still exists.  (which means the exception handling is occuring one layer lower than it should to me)

MRamonLeon committed:
[JENKINS-54459] Fix updateSiteRightJsonTest in Windows

MRamonLeon committed:
[JENKINS-54459] Minor fixes

jvz commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -1643,26 +1649,73 @@ public HttpResponse doUploadPlugin(StaplerRequest req) throws IOException, Servl
+        FormValidation response = null;
+        int i = 0;
+        boolean success = false;
+        while(i < retries && !success) {
```
Can't you combine the two variables into a for loop?

MRamonLeon replied:
As the two variables have different types and cannot be defined in the initialization section of the for bucle, I prefer to leave it so

jvz commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -1643,26 +1649,73 @@ public HttpResponse doUploadPlugin(StaplerRequest req) throws IOException, Servl
+    }
+
+    private FormValidation checkUpdatesServer() throws Exception {
+        for (UpdateSite site : Jenkins.getInstance().getUpdateCenter().getSites()) {
```
```suggestion
        for (UpdateSite site : Jenkins.get().getUpdateCenter().getSites()) {
```

jvz committed:
Avoid deprecated getInstance

Co-Authored-By: MRamonLeon <manuelramonleonjimenez@gmail.com>

Wadeck commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -1643,26 +1649,73 @@ public HttpResponse doUploadPlugin(StaplerRequest req) throws IOException, Servl
+                    }
+                } else {
+                    LOGGER.info(String.format("Attempted to check updates server for %s times with no success, raising an error", retries));
+                    LOGGER.severe(String.format("Error checking updates server for %s times. Last exception was: %s", retries, lastErrorCheckUpdateCenters));
```
Is that required to have two log with different level there ?

Wadeck commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -1643,26 +1649,73 @@ public HttpResponse doUploadPlugin(StaplerRequest req) throws IOException, Servl
+                } else {
+                    LOGGER.info(String.format("Attempted to check updates server for %s times with no success, raising an error", retries));
+                    LOGGER.severe(String.format("Error checking updates server for %s times. Last exception was: %s", retries, lastErrorCheckUpdateCenters));
+                    if(!LOGGER.isLoggable(Level.WARNING)) {
```
nit:

```suggestion
                    if (!LOGGER.isLoggable(Level.WARNING)) {
```

(add a space + test suggestion ;) )

Wadeck commented on test/src/test/java/hudson/model/UpdateCenterErrorsTest.java:
```diff
@@ -0,0 +1,185 @@
+    @Test
+    public void updateSiteRightJsonTest() throws Exception {
+        //Avoid CertPathValidatorException: Algorithm constraints check failed on signature algorithm: MD5withRSA
+        DownloadService.signatureCheck = false;
```
Perhaps worth using try{...} finally{...} to put back the initial value after that as it could have side effect?

(no suggestion as it modifies the whole code in the method ^^)

Wadeck replied:
{{@Before}} / {{@After}} are global for the tests but those configurations seem to be related to this specific one.

To be more precise about my proposal:
````java
boolean currentValue = DownloadService.signatureCheck;
try{ 
   DownloadService.signatureCheck = false;
   checkUpdateSite(Jenkins.getInstance().getRootUrl() + "updateSiteRightJson/update-center.json", "", true );
}
finally{
   DownloadService.signatureCheck = currentValue;
}
```

With that approach we do not let side effects in place after the run. (not sure it has any)

Wadeck commented on test/src/test/java/hudson/model/UpdateCenterErrorsTest.java:
```diff
@@ -0,0 +1,185 @@
+     * @throws SAXException If an exception is thrown using the UI.
+     */
+    private void checkUpdateSite(String urlUpdateSite, String message, boolean isSuccess) throws IOException, SAXException {
+        if(urlUpdateSite != null) {
```
nit:
```suggestion
        if (urlUpdateSite != null) {
```

varyvol replied:
Actually I think it can be removed, since it's never null?

Wadeck commented on test/src/test/java/hudson/model/UpdateCenterErrorsTest.java:
```diff
@@ -0,0 +1,185 @@
+        Page pageAfterClick = HtmlElementUtil.click(getCheckNow(p));
+        String page = pageAfterClick.getWebResponse().getContentAsString();
+
+        if(isSuccess)
```
nit: add `{...}` (and spaces ^^)
or put `assertEquals(isSuccess, page.contais...)` to simplify ?

varyvol replied:
+1

fcojfernandez replied:
+1

batmat replied:
Just a meta-comment: can we maybe use the embedded "Pick your reaction" to add _+1_ and the likes, so that we 1) get less notifications and 2) the PR grows a _bit_ less in height? Thanks :) :hugs:

Wadeck commented on test/src/test/java/hudson/model/UpdateCenterErrorsTest.java:
```diff
@@ -0,0 +1,185 @@
+
+        @Override
+        public @CheckForNull
+        String getIconFileName() {
```
nit: anormal formatting (IDE auto?) 
same for second one

Wadeck commented on test/src/test/java/hudson/model/UpdateCenterErrorsTest.java:
```diff
@@ -0,0 +1,185 @@
+        }
+
+        // The url has to end in update-center.json. See: UpdateSite#getMetadataUrlForDownloadable
+        public void  doDynamic(StaplerRequest staplerRequest, StaplerResponse staplerResponse) throws ServletException, IOException {
```
nit: (removal of a double spaces)

```suggestion
        public void doDynamic(StaplerRequest staplerRequest, StaplerResponse staplerResponse) throws ServletException, IOException {
```

Wadeck commented on test/src/test/java/hudson/model/UpdateCenterErrorsTest.java:
```diff
@@ -0,0 +1,185 @@
+import static org.junit.Assert.assertEquals;
+
+public class UpdateCenterErrorsTest {
+    private static final Logger LOGGER = Logger.getLogger(Jenkins.class.getName());
```
```suggestion
    private static final Logger LOGGER = Logger.getLogger(UpdateCenterErrorsTest.class.getName());
```

Perhaps easier to understand (otherwise we could ask why using Jenkins there)

Wadeck commented on test/src/test/java/hudson/model/UpdateCenterErrorsTest.java:
```diff
@@ -0,0 +1,185 @@
+                }
+            };
+
+            return r;
```
You could write something simpler like:

```java
public void doGetJson(StaplerRequest request, StaplerResponse response) {
   response.setContentType("text/json");
   response.setStatus(200);
   response.getWriter().append("{wrongjson}"");
}
```

jeffret-b commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -1643,26 +1649,73 @@ public HttpResponse doUploadPlugin(StaplerRequest req) throws IOException, Servl
-                }
+
+        int sleep = SystemProperties.getInteger(PluginManager.class.getName() + ".sleepTimeMilis", 1000);
+        int retries = SystemProperties.getInteger(PluginManager.class.getName() + ".retries", 3);
```
It might be clearer if it indicated what operation these variables relate to. Maybe something like "checkUpdatesSleepTimeMillis" and "checkUpdatesRetries".

varyvol replied:
+1

fcojfernandez replied:
@MRamonLeon, @batmat is right. In general this should be read in a static block when the class is initialized

jeffret-b commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -1643,26 +1649,73 @@ public HttpResponse doUploadPlugin(StaplerRequest req) throws IOException, Servl
-                    return v;
+
+            if (!success) {
+                LOGGER.warning(String.format("The attempt #%s to check updates server failed. Exception: %s", i, lastErrorCheckUpdateCenters));
```
I think it reads a little easier to change "updates" to "update".

jeffret-b replied:
And the same in a number of other places in this PR.

jeffret-b reviewed (approved):
Looks good.

fcojfernandez commented on core/src/main/resources/hudson/Messages_es.properties:
```diff
@@ -55,3 +55,4 @@ PluginManager.PluginDoesntSupportDynamicLoad.RestartRequired=El plugin {0} no so
 PluginManager.PortNotANumber=El puerto no es un n�mero
 AboutJenkins.Description=Eche un vistazo a la informaci�n sobre la versi�n y la licencia.
 PluginManager.PortNotInRange=El puerto no est� en el rango de {0} hasta {1}
+PluginManager.CheckUpdateServerError=Hubo errores comprobando los servidores de actualizaciones: {0}
```
I know it's out the scope of this PR, but what about to complete the translation of the messages file? It would be nice since you have added your message :)

batmat replied:
I agree about the usefulness, while at it, *but* I'm strongly against adding such commit in this PR. Keep PRs small, file many if needed, we all win :-). Shorter reviews, more chance to get merged and not block oneself for an unrelated change that someone would disagree on.

fcojfernandez replied:
An I agree with you @batmat in general and obviously if it would be a reformat of the code. However, in this case it's just text in his native language and he's modifying the file, that's why I'm suggesting. If he shouldn't have to add his message, my vote would be absolutely for not changing the content of the file.
IMO, if we leave that sort of changes for other PRs, the messages will likely keep out-of-date, I'm afraid :disappointed: 

But if everyone else that review the PR think it's better to open a new PR, it's fine for me.

fcojfernandez replied:
@batmat In most cases I actually agree with you. I just see this file a bit different just because it's only one of the resource boundles for a concrete message, but as I say before, I'm happy if it's done in another PR :)

fcojfernandez reviewed (approved):
Code looks good to me. Just minor comments

varyvol commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -1643,26 +1649,73 @@ public HttpResponse doUploadPlugin(StaplerRequest req) throws IOException, Servl
-                    return v;
-                }
+
+        int sleep = SystemProperties.getInteger(PluginManager.class.getName() + ".sleepTimeMilis", 1000);
```
I'm not sure about defaulting to 3 retries in case the property is not specified. Maybe we should default to no retry (as of now) and retry only in case the property is specified.

fcojfernandez replied:
I like this approach, @varyvol

varyvol commented on core/src/main/resources/hudson/PluginManager/check.jelly:
```diff
@@ -29,5 +29,9 @@ THE SOFTWARE.
   </f:link>
-
+  <j:if test="${app.pluginManager.lastErrorCheckUpdateCenters != null}">
+    <div class="error">
```
Maybe some kind of heading explaining that the error happened when checking the update centers?

varyvol commented on test/src/test/java/hudson/model/UpdateCenterErrorsTest.java:
```diff
@@ -0,0 +1,185 @@
+        public void  doDynamic(StaplerRequest staplerRequest, StaplerResponse staplerResponse) throws ServletException, IOException {
+            staplerResponse.setContentType("text/json");
+            staplerResponse.setStatus(200);
+            staplerResponse.serveFile(staplerRequest, UpdateCenterErrorsTest.class.getResource("update-center.json"));
```
This file already existed?

batmat commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -1643,26 +1649,73 @@ public HttpResponse doUploadPlugin(StaplerRequest req) throws IOException, Servl
-                if (v.kind != FormValidation.Kind.OK) {
-                    return v;
+
+            if (!success) {
```
not sure the role of `success` is required? Couldn't you move all this above in the try block, and finish with a `break;` to just go out of the loop in case of success?

batmat reviewed (changes requested):
Awesome change! Long time needed, thanks for it! :-)

I left a few comments, not all are required to be addressed per code.

I would like to see the following:
* rename `retries` to `attempts` if my comment below is right, and `retries=3` would actually try 3 times overall, which would be wrong.
* Extract that added logic into a separate method or class, so that it's easier to be tested separately and understood.

Thanks!!!

batmat commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -1643,26 +1649,73 @@ public HttpResponse doUploadPlugin(StaplerRequest req) throws IOException, Servl
+        retries = retries > 0 ? retries : 3;
+        sleep = sleep > 0 ? sleep : 1000;
+
+        FormValidation response = null;
```
BTW you can move/inline this below I think, no need to keep out of the sole location below where it's used.

batmat reviewed (commented):
a few more comments I had forgotten to publish

fcojfernandez commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -1643,26 +1649,73 @@ public HttpResponse doUploadPlugin(StaplerRequest req) throws IOException, Servl
+        int sleep = SystemProperties.getInteger(PluginManager.class.getName() + ".sleepTimeMilis", 1000);
+        int retries = SystemProperties.getInteger(PluginManager.class.getName() + ".retries", 3);
+
+        retries = retries > 0 ? retries : 3;
```
@batmat the 0 retry is what @varyvol is propossing in his comment above.

fcojfernandez commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -1643,26 +1649,73 @@ public HttpResponse doUploadPlugin(StaplerRequest req) throws IOException, Servl
+        int retries = SystemProperties.getInteger(PluginManager.class.getName() + ".retries", 3);
+
+        retries = retries > 0 ? retries : 3;
+        sleep = sleep > 0 ? sleep : 1000;
```
@batmat do you mean avoiding both system properties or just this one? I think maybe the time to suspend the instance shouldn't be opened to any value, but I don't see any problem in letting the user to configure the number or attempts if we just control and validate that the value is always a positive integer.

batmat replied:
@fcojfernandez I mean we shouldn't care to control is the value is actually an integer. I understand the goal here, but I'm saying sysprops are **very*** dangerous by nature, and we have a bunch of them already that I know we're not checking, and you will often just break the whole system if you put the wrong value. I guess this is ok here. I'm just trying to be aggressive with not adding more lines than we need to, ideally even remove some overall when possible :).

MRamonLeon committed:
[JENKINS-54459] Some fixes

- Better system properties initialization
- Rename properties
- Localized messages
- Add test for change log level message
- Restore signatureCheck field after test
- Other minor improvements

MRamonLeon committed:
[JENKINS-54459] Add Retrier generic class and use it in the checking

MRamonLeon reviewed (commented):
Changed code as requested from your reviews and also changed the _retry procedure_ to be accomplished in a new generic class (```hudson.util.Retrier```) available to whoever wants to do a _try-retry_ procedure with any action.

MRamonLeon committed:
[JENKINS-54459] Improving Spanish messages

MRamonLeon committed:
Merge remote-tracking branch 'origin/JENKINS-54459' into JENKINS-54459

MRamonLeon committed:
[JENKINS-54459] merge with remote master

varyvol commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -177,6 +178,31 @@
+    /**
+     * Time elapsed between retries to check the updates sites.
+     */
+    public static int checkUpdateSleepTimeMillis;
```
Why public?

Wadeck replied:
As this "variable" is meant to be a constant, we can name it `CHECK_UPDATE_SLEEP_TIME_MILLIS`. The `public` could be removed but please do not add `final` or `private`.

1) The `private` will prevent us to modify it in tests, to ensure its modification is effective, otherwise its behavior cannot be tested (without reflection I mean).
2) The `final` will prevent Script Console to modify its value on the fly (Groovy does not care about `private` but consider the `final`)
3) Yes you can have a `static final` variable that is not yet initialized but only in a subsequent static block.

varyvol commented on core/src/main/java/hudson/util/Retrier.java:
```diff
@@ -0,0 +1,177 @@
+        private String action;
+        private BiPredicate<Integer, V> checkResult;
+
+        private int attempts = 3;
```
I think 1 would be a better default value?

Wadeck commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -1643,26 +1649,73 @@ public HttpResponse doUploadPlugin(StaplerRequest req) throws IOException, Servl
+                    try {
+                        Thread.sleep(sleep);
+                    } catch (InterruptedException ie) {
+                        LOGGER.info("The wait for a new attempt to check updates server was interrupted. Exception: " + ie);
```
What he proposed I think is more something like:

```suggestion
                        LOGGER.info("The wait for a new attempt to check updates server was interrupted.", ie);
```

MRamonLeon committed:
[JENKINS-54459] Fix two javadoc errors

Wadeck commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -177,6 +178,31 @@
+            // Secure initialization
+            checkUpdateSleepTimeMillis = SystemProperties.getInteger(PluginManager.class.getName() + ".checkUpdateSleepTimeMillis", 1000);
+            checkUpdateAttempts = SystemProperties.getInteger(PluginManager.class.getName() + ".checkUpdateAttempts", 1);
+        } catch(Exception e) {
```
As you are passing default value, there is no reason to catch an exception there. The only problem that could arise is a `null` returned by the method but with the default value, there is no risk. That could simplify a bit the code there.

Wadeck commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -1654,24 +1686,86 @@ public HttpResponse doUploadPlugin(StaplerRequest req) throws IOException, Servl
+            // Check how it went
+            if (!FormValidation.Kind.OK.equals(result.kind)) {
+                LOGGER.log(Level.SEVERE, Messages.PluginManager_UpdateSiteError(checkUpdateAttempts, result.getMessage()));
+                if (checkUpdateAttempts > 1 && !Logger.getLogger(Retrier.class.getName()).isLoggable(Level.WARNING)) {
```
Due to previous log message + this weird call, I am not sure that if block is necessary.

fcojfernandez replied:
I think this second message can be unclear. @MRamonLeon  could you share an example about how it is shown and clarify the purpose of the message? Thanks!

MRamonLeon committed:
[JENKINS-54459] Rename constants and its visibility

Rename the constants to do the retries and change their visibilities.

Change the package of the test to be able to access the constants in
order to redefine their values in the tests.

MRamonLeon reviewed (commented):
Rename in _constants_ and their visibility

batmat commented on core/src/test/java/hudson/util/RetrierTest.java:
```diff
@@ -0,0 +1,260 @@
+        assertTrue(String.format("The log should contain '%s'", textNoSuccess), handler.getView().stream().anyMatch(m -> m.getMessage().contains(textNoSuccess)));
+
+        String testException = Messages.Retrier_ExceptionFailed(ATTEMPTS, ACTION);
+        assertTrue(String.format("The log should contain '%s'", testException), handler.getView().stream().anyMatch(m -> m.getMessage().startsWith(testException)));
```
Possibly, you could add an assertion checking that between the `build()` and here, there was at least `2x100` milliseconds elapsed? (100 after the first and the second attempt).

MRamonLeon replied:
I've added a new test to check the sleep time works well

batmat commented on core/src/main/java/hudson/util/Retrier.java:
```diff
@@ -0,0 +1,177 @@
+ * listener to manage the expected exceptions happened, just in case it is helpful.
+ * @param <V> The return type of the action to perform.
+ */
+public class Retrier <V>{
```
I think we should mark this `@Restricted(NoExternalUse.class)` at least for some time to let this soak, so that we can possibly fix/improve the design in future releases before it is possibly relaxed and made available to all plugins out there?

batmat reviewed (approved):
Nice!
I think the `Retrier` class is going to be pretty useful. 

I do think we should mark the class `@Restricted(NoExternalUse.class)` for some time, but I'll leave you and others see if this is a good idea. Do not want to block this more.

MRamonLeon committed:
[JENKINS-54459] Avoid external use of Retrier and test the delay time

MRamonLeon reviewed (commented):
Added the _Restricted_ annotation until it is used enough ( :thinking: when?) and added an extra test to check the delay time really works.

batmat commented:
Given the time since this PR has been opened, I think we can target merging it before the end of the week. Possibly some time tomorrow or Friday.

I'll do it if nobody objects in the meantime.

Thanks a lot Ramon!

batmat commented:
> Added the Restricted annotation until it is used enough (when)

No strict rule. I'd say at least the current PR is battle tested a few weeks, and possibly/hopefully we can use it elsewhere in the Core and check it's not just usable in the context here. Then we'd open it _to the world_ I guess.

oleg-nenashev reviewed (commented):
The code itself looks good to me. My main concern is that the `Retrier` class looks to be designed for a single case, and it may be producing a lot of overhead if it used in other cases. Have you considered creating an implementation based on https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/model/AperiodicWork.java instead of creating a new one?

oleg-nenashev reviewed (approved):
@batmat @MRamonLeon  I do not consider my comments as blocker ones, because the performance impact of update center should be really limited. I will be happy to merge it, assuming that there will be a follow-up task to cleanup the implementation

batmat commented:
:+1: Will merge it as soon as the PR build is green. Feel free to do if you come here before I do next time :).
